pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,337 @@
+			AccessModes:      pvc.Spec.AccessModes,
+			Resources:        pvc.Spec.Resources,
+			StorageClassName: pvc.Spec.StorageClassName,
+			VolumeMode:       pvc.Spec.VolumeMode,
+		},
+	}
+	pvcPrime.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(pvc, schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "PersistentVolumeClaim",
+		}),
+	}
+	util.SetRecommendedLabels(pvcPrime, r.installerLabels, "cdi-controller")
+
+	if !isPVCOwnedByDataVolume(pvc) {
+		requestedSize, _ := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		// disk or image size, inflate it with overhead
+		requestedSize, err := cc.InflateSizeWithOverhead(r.client, requestedSize.Value(), &pvc.Spec)
+		if err != nil {
+			return nil, err
+		}
+		pvcPrime.Spec.Resources.Requests[corev1.ResourceStorage] = requestedSize
+
+	}
+	if pvcModifier != nil {
+		pvcModifier(pvcPrime, source)
+	}
+	if err := r.client.Create(context.TODO(), pvcPrime); err != nil {
+		r.recorder.Eventf(pvc, corev1.EventTypeWarning, errCreatingPVCPrime, err.Error())
+		return nil, err
+	}
+
+	r.recorder.Eventf(pvc, corev1.EventTypeNormal, createdPVCPrimeSuccessfully, createdPVCPrimeSuccessfully)
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) rebindPV(targetPVC, pvcPrime *corev1.PersistentVolumeClaim) error {
+	pv, err := r.getPV(pvcPrime.Spec.VolumeName)
+	if pv == nil {
+		return err
+	}
+
+	// Examine the claimref for the PV and see if it's still bound to PVC'
+	claimRef := pv.Spec.ClaimRef
+	if claimRef.Name != pvcPrime.Name || claimRef.Namespace != pvcPrime.Namespace || claimRef.UID != pvcPrime.UID {
+		return nil
+	}
+
+	// Rebind PVC to target PVC
+	pv.Annotations = make(map[string]string)
+	pv.Spec.ClaimRef = &corev1.ObjectReference{
+		Namespace:       targetPVC.Namespace,
+		Name:            targetPVC.Name,
+		UID:             targetPVC.UID,
+		ResourceVersion: targetPVC.ResourceVersion,
+	}
+	r.log.V(3).Info("Rebinding PV to target PVC", "PVC", targetPVC.Name)
+	if err := r.client.Update(context.TODO(), pv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcilerBase) reconcile(req reconcile.Request, populator populatorController, log logr.Logger) (reconcile.Result, error) {
+	// Get the target PVC
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// We first perform the common reconcile steps.
+	// We should only continue if we get a valid PVC'
+	pvcPrime, err := r.reconcileCommon(pvc, populator, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if pvcPrime == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Each populator reconciles the target PVC in a different way
+	if cc.IsUnbound(pvc) {
+		return populator.reconcileTargetPVC(pvc, pvcPrime)
+	}
+
+	// Making sure to clean PVC'
+	return r.reconcileCleanup(pvcPrime)
+}
```

Comment:
maybe I missed it, but I want to make sure we are handling the case where the target pvc has deletion timestamp set.  We shouldn't create PVC' in that case.  But maybe no need to call cleanup if everything has has ownerrefs

Reply:
Any recomendation on how to handle this? Isn't enough to just check for pvc.DeletionTimestamp just after getting the pvc? For example:

```go
	// Get the target PVC
	pvc := &corev1.PersistentVolumeClaim{}
	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
		if k8serrors.IsNotFound(err) {
			return reconcile.Result{}, nil
		}
		return reconcile.Result{}, err
	}

        if pvc.DeletionTimestamp != nil {
                log.V(1).Info("PVC being terminated, ignoring it")
                return reconcile.Result{}, nil
        }
```

Reply:
I think with the ownerRef then the suggestion is good